src/mlpack/tests/dictionary_encoding_test.cpp

```diff
@@ -0,0 +1,179 @@
+}
+
+/**
+* Test for SplitByChar class.
+*/
+BOOST_AUTO_TEST_CASE(SplitByCharTest)
+{
+  std::vector<boost::string_view> tokens;
+  boost::string_view strview = "hello how are you";
+  SplitByChar obj(" ");
+  boost::string_view token;
+  token = obj(strview);
+  while (!token.empty())
+  {
+    tokens.push_back(token);
+    token = obj(strview);
+  }
+  BOOST_REQUIRE_EQUAL(tokens[0], "hello");
+  BOOST_REQUIRE_EQUAL(tokens[1], "how");
+  BOOST_REQUIRE_EQUAL(tokens[2], "are");
+  BOOST_REQUIRE_EQUAL(tokens[3], "you");
+}
+
+/**
+* Test Dictionary encoding for characters using lamda function.
+*/
+BOOST_AUTO_TEST_CASE(DictionaryEncodingCharTest)
+{
+  std::vector<string>arr(3);
+  arr[0] = "GACCA";
+  arr[1] = "ABCABCD";
+  arr[2] = "GAB";
+  arma::mat output;
+  data::DictionaryEncoding en;
+  // Passing a empty string to encode characters
+  en.Encode(arr, output, [](boost::string_view& str) {
+      if (str.empty())
+        return str;
+      boost::string_view retval = str.substr(0, 1);
+      str.remove_prefix(1);
+      return retval;
+  });
+  const std::unordered_map<boost::string_view, size_t,
+      boost::hash<boost::string_view>>& maps = en.Mappings();
+  // Checking that everything is mapped to different numbers.
+  std::unordered_map<size_t, size_t>cnt;
+  for (auto it = maps.begin(); it != maps.end(); it++)
+  {
+    cnt[it->second]++;
+    // Every token should be mappend only once.
+    BOOST_REQUIRE_EQUAL(cnt[it->second], 1);
+  }
+  arma::mat target = "1 2 3 3 2 0 0;"
+                     "2 4 3 2 4 3 5;"
+                     "1 2 4 0 0 0 0;";
+  CheckMatrices(output, target);
+}
+
+/**
+* Testing the functionality of copy constructor.
+*/
+BOOST_AUTO_TEST_CASE(CopyConstructorCharTest)
+{
+  std::vector<string>arr(3);
+  arr[0] = "hello how are you";
+  arr[1] = "i am good";
+  arr[2] = "Good how are you";
+  arma::sp_mat output;
+  data::DictionaryEncoding en;
+  en.Encode(arr, output, SplitByChar(" "));
+  const std::unordered_map<boost::string_view, size_t,
+      boost::hash<boost::string_view>>& maps = en.Mappings();
+  data::DictionaryEncoding en2 = en;
+  const std::unordered_map<boost::string_view, size_t,
+      boost::hash<boost::string_view>>& maps2 = en2.Mappings();
+  // Comparing both of them.
+  BOOST_REQUIRE_EQUAL(maps2 == maps, true);
+}
+
+/**
+* Test Dictionary encoding for output with no padding.
+*/
+BOOST_AUTO_TEST_CASE(DictionaryEncodingNoPaddingTest)
+{
+  std::vector<string>arr(2);
+  arr[0] = "GACCA";
+  arr[1] = "GAB";
+  data::DictionaryEncoding en;
+  std::vector<std::vector<size_t> > output;
+  en.Encode(arr, output, [](boost::string_view& str) {
+      if (str.empty())
+        return str;
+      boost::string_view retval = str.substr(0, 1);
+      str.remove_prefix(1);
+      return retval;
+  });
+  std::vector<std::vector<size_t>> req_output;
+  std::vector<size_t> temp = {1, 2, 3, 3, 2};
+  req_output.push_back(temp);
+  temp.clear();
+  temp = {1, 2, 4};
+  req_output.push_back(temp);
```

Comment:
Actually I meant the following:
```
std::vector<std::vector<size_t>> req_output = {
    {1, 2, 3, 3, 2},
    {1, 2, 4}
};
```

Reply:
oh that was very foolish of me !, I am still confused why didn't i write it that way . :) .